CLI patterns differ on Windows vs Linux

<!-- The bug you're experiencing might have already been reported!
Please search in the [issues](https://github.com/pinterest/ktlint/issues) before creating one. -->

## Expected Behavior
<!---Tell us what should happen. -->
<!-- If relevant, provide a link to the relevant section in the [Kotlin Coding Conventions](https://kotlinlang.org/docs/coding-conventions.html) or [Android Kotlin Style Guide](https://developer.android.com/kotlin/style-guide).  -->

<!-- Note: Ktlint should not produce any code which conflict with IntelliJ default code formatting. So please verify that your expectation is accepted by IntelliJ default code formatting. -->

The glob patterns as described in https://pinterest.github.io/ktlint/latest/install/cli/#globs should work the same way running ktlint on any underlying operating system and should match the same files given the same file structure.

## Observed Behavior
<!---Tell us what happens instead of the expected behavior -->
<!--- Provide the exact command which was executed but please -->
<!--- ensure that the flag "--log-level=debug" is added to the -->
<!--- command as well. Provide the output of this command. -->
At the moment (using 1.3.1) glob patterns do not match the same files on Windows as on Linux.
Here are two examples:
`*.kt` matches any file with the .kt file-ending  in any subdirectory on Windows, however the same glob does not match any subdirectory on Linux.

`**/test*/**` seems to drop / ignore the second slash on Windows, since it matches a file at the location "src/example/TestFile.kt". On Linux this behaviour can not be observed.

Generally Linux seems to be in line with the git manpage and online glob testing sites.
I suspect that the single star wildcard is not being handled correctly on Windows machines.

Sadly, the ktlint command does not produce any meaning full tracing output, similar to the test's output mentione below (with trace logging enabled):

```
[DEBUG] c.p.ktlint.cli.internal.FileUtils - Start walkFileTree from directory: 'C:\project\project1'
[DEBUG] c.p.ktlint.cli.internal.FileUtils - Discovered 2 files to be processed in 1 ms
[INFO] c.p.k.cli.internal.FileUtilsTest - Getting files with [patterns = [**/test*/**]] and [rootdir = C:\project\project1] returns [files = [project1/src/mock/kotlin/TestOneSetup.kt, project1/src/testMock/kotlin/TwoTest.kt]]
```

## Steps to Reproduce
<!--- Provide a code example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Minimize the example to the bare -->
<!--- minimum required to reproduce the bug. Small, well written -->
<!--- example are proven to be resolved quicker. -->

Following test can be added to https://github.com/pinterest/ktlint/blob/master/ktlint-cli/src/test/kotlin/com/pinterest/ktlint/cli/internal/FileUtilsTest.kt

```
    @Test
    fun `Find files with correct glob for wildcards`() {
        val ktFileInProjectOutsideFlavoredDirectoryWithOverlappingName = "project1/src/mock/kotlin/TestOneSetup.kt"
        val ktFileInProjectFlavoredDirectoryWithoutOverlappingName = "project1/src/testMock/kotlin/TwoTest.kt"

        ktlintTestFileSystem.createFile(ktFileInProjectOutsideFlavoredDirectoryWithOverlappingName)
        ktlintTestFileSystem.createFile(ktFileInProjectFlavoredDirectoryWithoutOverlappingName)

        val foundFiles =
            getFiles(
                patterns = listOf("**/test*/**"),
                rootDir = ktlintTestFileSystem.resolve(ktFileInProjectRootDirectory).parent.toAbsolutePath(),
            )

        assertThat(foundFiles)
            .containsExactlyInAnyOrder(
                ktFileInProjectFlavoredDirectoryWithoutOverlappingName
            ).doesNotContain(
                ktFileInProjectOutsideFlavoredDirectoryWithOverlappingName
            )
    }
```

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* Version of ktlint used: 1.3.1
* Issue occurs with or without any editorconfig being defined.
* Name and version (or code for custom task) of integration used (Gradle plugin, Maven plugin, command line, custom Gradle task): ktlint-cli
* Version of Gradle used (if applicable): none
* Operating System and version: observable on Windows 10 & Windows 11

Yes, that is correct. The test is failing on Windows:
```
java.lang.AssertionError: 
Expecting actual:
  ["project1/src/mock/kotlin/TestOneSetup.kt",
    "project1/src/testMock/kotlin/TwoTest.kt"]
to contain exactly in any order:
  ["project1/src/testMock/kotlin/TwoTest.kt"]
but the following elements were unexpected:
  ["project1/src/mock/kotlin/TestOneSetup.kt"]
```

--- a/ktlint-cli/src/main/kotlin/com/pinterest/ktlint/cli/internal/FileUtils.kt
+++ b/ktlint-cli/src/main/kotlin/com/pinterest/ktlint/cli/internal/FileUtils.kt
@@ -296,7 +296,7 @@ private fun String?.expandDoubleStarPatterns(): Set<String> {
             // trailing "**" matches any file in the directory.
             // Given pattern "**/Test*/**" the file "src/Foo/TestFoo.kt" should not be matched, and file "src/TestFoo/FooTest.kt" is
             // matched. If the original pattern would be expanded with additional pattern "**/Test*" then both files would have been
-            //  matched.
+            // matched.
             it === parts.last()
         }.forEach { doubleStarPart ->
             run {

--- a/ktlint-cli/src/main/kotlin/com/pinterest/ktlint/cli/internal/FileUtils.kt
+++ b/ktlint-cli/src/main/kotlin/com/pinterest/ktlint/cli/internal/FileUtils.kt
@@ -296,7 +296,7 @@ private fun String?.expandDoubleStarPatterns(): Set<String> {
             // trailing "**" matches any file in the directory.
             // Given pattern "**/Test*/**" the file "src/Foo/TestFoo.kt" should not be matched, and file "src/TestFoo/FooTest.kt" is
             // matched. If the original pattern would be expanded with additional pattern "**/Test*" then both files would have been
-            // matched.
+            //  matched.
             it === parts.last()
         }.forEach { doubleStarPart ->
             run {

--- a/ktlint-cli/src/main/kotlin/com/pinterest/ktlint/cli/internal/FileUtils.kt
+++ b/ktlint-cli/src/main/kotlin/com/pinterest/ktlint/cli/internal/FileUtils.kt
@@ -1,6 +1,11 @@
 package com.pinterest.ktlint.cli.internal
 
+import ch.qos.logback.classic.Level
+import ch.qos.logback.classic.Logger
 import com.pinterest.ktlint.logger.api.initKtLintKLogger
+import com.pinterest.ktlint.logger.api.setDefaultLoggerModifier
+import io.github.oshai.kotlinlogging.DelegatingKLogger
+import io.github.oshai.kotlinlogging.KLogger
 import io.github.oshai.kotlinlogging.KotlinLogging
 import org.jetbrains.kotlin.util.prefixIfNot
 import java.io.File
@@ -20,7 +25,22 @@ import kotlin.io.path.pathString
 import kotlin.io.path.relativeToOrSelf
 import kotlin.system.measureTimeMillis
 
-private val LOGGER = KotlinLogging.logger {}.initKtLintKLogger()
+private val LOGGER =
+    KotlinLogging
+        .logger {}
+        .setDefaultLoggerModifier { it.level = Level.TRACE }
+        .initKtLintKLogger()
+
+private var KLogger.level: Level?
+    get() = underlyingLogger()?.level
+    set(value) {
+        underlyingLogger()?.level = value
+    }
+
+private fun KLogger.underlyingLogger(): Logger? =
+    @Suppress("UNCHECKED_CAST")
+    (this as? DelegatingKLogger<Logger>)
+        ?.underlyingLogger
 
 private val ROOT_DIR_PATH: Path = Paths.get("").toAbsolutePath()
 
@@ -124,7 +144,7 @@ internal fun FileSystem.fileSequence(
                         if (negatedPathMatchers.none { it.matches(path) } &&
                             pathMatchers.any { it.matches(path) }
                         ) {
-                            LOGGER.trace { "- File: $path: Include" }
+                            LOGGER.trace { "- File: $path: Include as it matches patterns ${pathMatchers.filter { it.matches(path) }}" }
                             result.add(path)
                         } else {
                             LOGGER.trace { "- File: $path: Ignore" }
@@ -262,7 +282,7 @@ private fun FileSystem.toGlob(
 }
 
 /**
- * For each double star pattern in the path, create and additional path in which the double start pattern is removed.
+ * For each double star pattern in the path, create an additional path in which the double star pattern is removed.
  * In this way a pattern like some-directory/**/*.kt will match while files in some-directory or any of its
  * subdirectories.
  */
@@ -271,14 +291,21 @@ private fun String?.expandDoubleStarPatterns(): Set<String> {
     val parts = this?.split("/").orEmpty()
     parts
         .filter { it == "**" }
-        .forEach { doubleStarPart ->
+        .filterNot {
+            // When the pattern ends with a double star, it might not be expanded. According to the https://git-scm.com/docs/gitignore a
+            // trailing "**" matches any file in the directory.
+            // Given pattern "**/Test*/**" the file "src/Foo/TestFoo.kt" should not be matched, and file "src/TestFoo/FooTest.kt" is
+            // matched. If the original pattern would be expanded with additional pattern "**/Test*" then both files would have been
+            // matched.
+            it === parts.last()
+        }.forEach { doubleStarPart ->
             run {
+                // The original path can contain multiple double star patterns. Replace only one double star pattern
+                // with an additional path pattern and call recursively for remaining double star patterns
                 val expandedPath =
                     parts
                         .filter { it !== doubleStarPart }
                         .joinToString(separator = "/")
-                // The original path can contain multiple double star patterns. Replace only one double start pattern
-                // with an additional path pattern and call recursively for remain double star patterns
                 paths.addAll(expandedPath.expandDoubleStarPatterns())
             }
         }

--- a/ktlint-cli/src/test/kotlin/com/pinterest/ktlint/cli/internal/FileUtilsTest.kt
+++ b/ktlint-cli/src/test/kotlin/com/pinterest/ktlint/cli/internal/FileUtilsTest.kt
@@ -1,7 +1,12 @@
 package com.pinterest.ktlint.cli.internal
 
+import ch.qos.logback.classic.Level
+import ch.qos.logback.classic.Logger
 import com.pinterest.ktlint.logger.api.initKtLintKLogger
+import com.pinterest.ktlint.logger.api.setDefaultLoggerModifier
 import com.pinterest.ktlint.test.KtlintTestFileSystem
+import io.github.oshai.kotlinlogging.DelegatingKLogger
+import io.github.oshai.kotlinlogging.KLogger
 import io.github.oshai.kotlinlogging.KotlinLogging
 import org.assertj.core.api.Assertions.assertThat
 import org.junit.jupiter.api.AfterEach
@@ -15,7 +20,22 @@ import org.junit.jupiter.params.provider.ValueSource
 import java.nio.file.Path
 import kotlin.io.path.absolutePathString
 
-private val LOGGER = KotlinLogging.logger {}.initKtLintKLogger()
+private val LOGGER =
+    KotlinLogging
+        .logger {}
+        .setDefaultLoggerModifier { it.level = Level.TRACE }
+        .initKtLintKLogger()
+
+private var KLogger.level: Level?
+    get() = underlyingLogger()?.level
+    set(value) {
+        underlyingLogger()?.level = value
+    }
+
+private fun KLogger.underlyingLogger(): Logger? =
+    @Suppress("UNCHECKED_CAST")
+    (this as? DelegatingKLogger<Logger>)
+        ?.underlyingLogger
 
 /**
  * Tests for [fileSequence] method.
@@ -491,6 +511,25 @@ internal class FileUtilsTest {
                 ktFileRootDirectory,
                 ktsFileRootDirectory,
             )
+    }
+
+    @Test
+    fun `Issue 2781 - Given a glob containing a subdirectory matcher with wildcard, and ending with a double star then the subdirectory wildcard may not be used to match the filename`() {
+        val ktFileInProjectOutsideTargetedDirectory = "project1/src/TestFoo.kt"
+        val ktFileInProjectInsideTargetedDirectory = "project1/src/Foo/FooTest.kt"
+
+        ktlintTestFileSystem.createFile(ktFileInProjectOutsideTargetedDirectory)
+        ktlintTestFileSystem.createFile(ktFileInProjectInsideTargetedDirectory)
+
+        val foundFiles =
+            getFiles(
+                patterns = listOf("**/*Foo*/**"),
+                rootDir = ktlintTestFileSystem.resolve(ktFileInProjectRootDirectory).parent.toAbsolutePath(),
+            )
+
+        assertThat(foundFiles)
+            .containsExactlyInAnyOrder(ktFileInProjectInsideTargetedDirectory)
+            .doesNotContain(ktFileInProjectOutsideTargetedDirectory)
     }
 
     private fun KtlintTestFileSystem.createFile(fileName: String) =

